3.x fix compatibility with android talkback

chriskyfung opened the pull request:
## Description

The hamburger menu and its submenu cannot toggle on Android when using Android TalkBack to interact with the menu controller and submenu togglers.

TalkBack fires a mouse click event when users interact the menu togglers with double taps or choose an action from TalkBack's context menu. But the _handleClick() in `_baseMenu.js` solely add event listeners to response `touchend` and omit `mouseup`  events if v3.0.0 detects a touch screen device.

A fix was made by using for-loops to make sure the menu togglers will have both the `touchend` and `mouseup`  event listeners.

## Related Issues

[Violation] Added non-passive event listener to a scroll-blocking 'touchstart' event. Consider marking event handler as 'passive' to make the page more responsive. See https://www.chromestatus.com/feature/5745543795965952 (_baseMenu.js:937)

Added `{ passive : true }` as the option to the `touchstart` event listener.

chriskyfung committed:
fix(_basemenu.js): modified _handleClick() to support toggle togglers with Talkback on Android

The change allows handling click events triggered by Android TalkBack when activating controllers
or submenu togglers by double tapping or choosing an action from Talkback context menu.

chriskyfung committed:
perf(_basemenu.js): make event handler for focusChild() as 'passive'

A fix to the warning for _baseMenu.js:937 in Chrome DevTools: [Violation] Added non-passive event
listener to a scroll-blocking 'touchstart' event. Consider marking event handler as 'passive' to
make the page more responsive. See https://www.chromestatus.com/feature/5745543795965952

NickDJM commented:
Thanks for the PR!

The reason why we ommit the `mouseup` on touch devices was some browsers were seeing double events triggered. We'll need to make sure this isn't the case if we add both mouse/touch events to the toggles before this gets merged in.

chriskyfung commented:
@NickDJM Thanks for your feedback!

How about using [Pointer Events](https://developer.mozilla.org/en-US/docs/Web/API/Pointer_events) to replace both Touch and Mouse Events?

NickDJM commented:
@chriskyfung I like that idea a lot! It would definitely be a lot cleaner.

chriskyfung committed:
fix(_basemenu.js): go for pointer events in _handleClick() instead of adding both mouse/touch events

Pick `pointerdown` and `pointerup` as the default; Otherwise, fallback to either `touchstart` and
`touchend`  or `mousedown` and `mouseup` events as in v3.0.0

chriskyfung commented:
I revised the code and tested on my devices. It looks good in Chrome and Firefox on Windows as well as Chrome with Talkback on Android.

NickDJM commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "accessible-menu",
-  "version": "3.0.0",
+  "version": "3.0.1-alpha",
```
Don't change the version. The release script will take care of this.

chriskyfung replied:
Sorry for that! I will restore it in the next commit.

NickDJM commented on src/_baseMenu.js:
```diff
@@ -889,28 +889,32 @@ class BaseMenu {
   _handleClick() {
     // Use touch over mouse events when supported.
-    const startEventType = isEventSupported("touchstart", this.dom.menu)
+    const startEventType = isEventSupported("pointerdown", this.dom.menu)
```
Given that all supported browsers support `pointerup`/`pointerdown` events we should clean this section up to only use the pointer events, this will get rid of the need for the event checks.

Doc comment will need to be updated as well.

chriskyfung replied:
I kept the fallback to the original approach because I was not sure whether you just consider the last 2 versions of the supported browsers.

I can amend it if you confirm that is unnecessary to support old browser versions.

NickDJM replied:
I _believe_ we're in the clear for all the browsers that are supported. Chrome/FF/Edge have had full support for pointer events for a while now and Safari has had support since 2 versions ago.

chriskyfung replied:
I've removed the event checks and updated the comment. Please review the changes

chriskyfung replied:
Menu toggles and the `aria-expanded` attribute changes when I tested it in real browsers.

I think the test failed because JSDOM hasn't support PointEvent yet.

**References**:

- [PointerEvent constructor not available in window · Issue #2527 · jsdom/jsdom](https://github.com/jsdom/jsdom/issues/2527)
- [tc15-PointerEvent-not-available · tmobile/jest-jsdom-browser-compatibility](https://github.com/tmobile/jest-jsdom-browser-compatibility/blob/master/testcases-src/tc15-PointerEvent-not-available/README.md)

NickDJM reviewed (changes requested):
@chriskyfung Took a quick look through the changes and made some comments.

I'll also be doing a quick browser compatibility test (mainly for Safari on MacOS/iOS) just to make sure they still function as expected.

chriskyfung committed:
Merge branch '3.x' into 3.x-fix-compatibility-with-android-talkback

chriskyfung committed:
style(package.json): restore the version to v3.0.0

chriskyfung committed:
refactor(_basemenu.js): remove the event checks for pointerup/pointerdown events in _handleClick()

chriskyfung committed:
fix(jest test environment): bump jsdom from 16.6.0 to 18.0.0 via jest-environment-jsdom-latest

Fix Point Events not supported by JSDOM v16.6.0.

chriskyfung committed:
fix(test): stimulate `pointerdown` and `pointerup` instead of `mousedown` and `mouseup` events

1. Added PointerEvent class by extending jsdom.MouseEvent;
2. Added simulatePointerEvent() to simulate a pointer event;
3. Replaced simulateMouseEvent() with simulatePointerEvent() in simulateclick().

chriskyfung committed:
fix(_basemenu.js): change to add `pointerup` event lisenters for toggleToggle by addEventListenser()

Fix `pointerup` event callback not execute in Jest if register the callback by using `onpointerup`
attribute.

chriskyfung committed:
fix(_handleclick()): change to use `pointerup` listener for closing menu when user clicks outside it

Removed the `touchend`/`mouseup` event checks and replaced the with `pointerup` in menubar.js and
disclosureMenu.js

NickDJM commented:
Tests have passed. I'll do a code review and browser (Safari) tests tonight.

NickDJM commented on tests/menus/_common/helpers.js:
```diff
@@ -66,14 +102,14 @@ export function simulateKeyboardEvent(eventType, element, options = {}) {
 
 /**
- * Simulates a mousedown and mouseup event on a DOM element.
+ * Simulates a pointer event on a DOM element.
```
```suggestion
 * Simulates a pointerdown and a pointerup event on a DOM element.
```

NickDJM commented on src/menubar.js:
```diff
@@ -151,23 +150,18 @@ class Menubar extends BaseMenu {
-    const endEventType = isEventSupported("touchend", this.dom.menu)
-      ? "touchend"
-      : "mouseup";
+    const endEventType = "pointerup";
```
Same as other instance above. Lets just use "pointerup" in the event handler instead of declaring a variable.
```suggestion
```

NickDJM commented on src/menubar.js:
```diff
@@ -151,23 +150,18 @@ class Menubar extends BaseMenu {
    *
-   * Depending on what is supported either `touchstart` and `touchend` or
-   * `mousedown` and `mouseup` will be used for all "click" event handling.
+   * `pointerup` will be used for all "click" event handling.
```
Don't need this line anymore.
```suggestion
```

NickDJM commented on src/disclosureMenu.js:
```diff
@@ -204,23 +204,18 @@ class DisclosureMenu extends BaseMenu {
-    const endEventType = isEventSupported("touchend", this.dom.menu)
-      ? "touchend"
-      : "mouseup";
+    const endEventType = "pointerup";
```
Let's just get rid of this `const` since it's only used in one place and there is no longer any logic to figure which event to use.
```suggestion
```

NickDJM commented on src/disclosureMenu.js:
```diff
@@ -204,23 +204,18 @@ class DisclosureMenu extends BaseMenu {
    *
-   * Depending on what is supported either `touchstart` and `touchend` or
-   * `mousedown` and `mouseup` will be used for all "click" event handling.
+   * `pointerup` will be used for all "click" event handling.
```
Don't need this line anymore.

```suggestion
```

NickDJM commented on src/_baseMenu.js:
```diff
@@ -934,27 +927,37 @@ class BaseMenu {
-        this.focusChild(index);
-      });
+      item.dom.link.addEventListener(
+        startEventType,
```
```suggestion
        "pointerdown",
```

NickDJM commented on src/menubar.js:
```diff
@@ -151,23 +150,18 @@ class Menubar extends BaseMenu {
+    const endEventType = "pointerup";
 
     // Close the menu if a click event happens outside of it.
     document.addEventListener(endEventType, (event) => {
```
```suggestion
    document.addEventListener("pointerup", (event) => {
```

NickDJM commented on src/disclosureMenu.js:
```diff
@@ -204,23 +204,18 @@ class DisclosureMenu extends BaseMenu {
+    const endEventType = "pointerup";
 
     // Close the menu if a click event happens outside of it.
     document.addEventListener(endEventType, (event) => {
```
```suggestion
    document.addEventListener("pointerup", (event) => {
```

NickDJM commented on src/_baseMenu.js:
```diff
@@ -934,27 +927,37 @@ class BaseMenu {
-          toggleToggle(this, item.elements.toggle, event);
-        };
+        item.elements.toggle.dom.toggle.addEventListener(
+          endEventType,
```
```suggestion
          "pointerup",
```

NickDJM commented on src/_baseMenu.js:
```diff
@@ -934,27 +927,37 @@ class BaseMenu {
-        toggleToggle(this, this.elements.controller, event);
-      };
+      this.elements.controller.dom.toggle.addEventListener(
+        endEventType,
```
```suggestion
        "pointerup",
```

NickDJM commented on src/_baseMenu.js:
```diff
@@ -889,30 +888,24 @@ class BaseMenu {
-    const endEventType = isEventSupported("touchend", this.dom.menu)
-      ? "touchend"
-      : "mouseup";
+    const startEventType = "pointerdown";
+    const endEventType = "pointerup";
```
Since there's no longer any logic behind these we'll just use the strings in the events themselves.

```suggestion
```

NickDJM reviewed (commented):
@chriskyfung Reviewed with suggestions. Going to apply them and then run some final tests. This looks good!

NickDJM commented on src/_baseMenu.js:
```diff
@@ -889,30 +888,24 @@ class BaseMenu {
    *
-   * Depending on what is supported either `touchstart` and `touchend` or
-   * `mousedown` and `mouseup` will be used for all "click" event handling.
+   * `pointerdown` and `pointerup` will be used for all "click" event handling.
```
```suggestion
```

NickDJM committed:
chore: apply suggestions from review
